### PR TITLE
Support step and list attributes on DateTime picker

### LINKED ISSUE #1257: Support `step` and `list` attributes on DateTime picker
Support `step` attribute  and `list` attribute in `<input type="date|time|month|week|datetime-local">` on DateTime picker 

HTML Spec : https://html.spec.whatwg.org/multipage/input.html#date-state-(type=date)

Demo :
for `step` : https://www.w3schools.com/jsref/tryit.asp?filename=tryjsref_input_date_step
for `list` : https://demo.agektmr.com/datalist/

### PR DESCRIPTION
**Pull request by ojuschugh1**
Handle the HTML step attribute on date/time inputs by snapping the selected value to the nearest valid step boundary when the user confirms. For Chromium, datalist suggestions are also forwarded and shown in a Spinner above the pickers so the user can pick from them.

On the Gecko side step is read from GeckoView directly but datalist is not exposed by the engine so we just fall back gracefully.

Fixes #1257

### CODE REVIEW

**Review by svillar — CHANGES_REQUESTED** [review]
Looking good. I have a few comments

**Comment by svillar** [comment]
@ojuschugh1 are you planing to work on this, or should we just take over?

**Comment by ojuschugh1** [comment]
> @ojuschugh1 are you planing to work on this, or should we just take over?

Hi @svillar  sure i will look into this. I was on vacation apologies for late reply.

**Comment by svillar** [comment]
> > @ojuschugh1 are you planing to work on this, or should we just take over?
> 
> Hi @svillar sure i will look into this. I was on vacation apologies for late reply.

Awesome thanks, I hope you don't feel pressured, it's just this would be a nice to have for the upcoming release

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/PromptDelegateImpl.java [hunk 1]
```diff
@@ -699,6 +705,42 @@ public void setDateTime(int type, double value, double min, double max) {
...
+            // time).  Normalise it back to the raw HTML attribute unit so the
+            // widget receives the same value as the Gecko backend.
+            if (!Double.isNaN(step) && step > 0) {
+                if (mType == Type.DATE) {
+                    step = step / 86400000.0;
+                } else if (mType == Type.TIME || mType == Type.DATETIME_LOCAL) {
+                    step = step / 1000.0;
+                } else if (mType == Type.WEEK) {
+                    step = step / 604800000.0;
+                }
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/PromptDelegateImpl.java:724]: Done, extracted them to MS_PER_SECOND, MS_PER_DAY, MS_PER_WEEK constants. Applied in both PromptDelegateImpl and DateTimePromptWidget.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/PromptDelegateImpl.java [hunk 2]
```diff
@@ -699,6 +705,42 @@ public void setDateTime(int type, double value, double min, double max) {
...
+                List<String> listVals = new ArrayList<>();
+                for (DateTimeSuggestion s : suggestions) {
+                    try {
+                        double sv = s.value();
+                        if (mType == Type.MONTH) {
+                            sv = MonthPicker.createDateFromValue(sv).getTimeInMillis();
+                        } else if (mType == Type.WEEK) {
+                            sv = WeekPicker.createDateFromValue(sv).getTimeInMillis();
+                        }
+                        listVals.add(mFormatter.format(new Date((long) sv)));
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/PromptDelegateImpl.java:742]: That's actually intentional - for DATE, TIME, and DATETIME_LOCAL the suggestion value from Chromium is already in milliseconds since epoch, so it can go straight into new Date((long) sv). Only MONTH and WEEK use a special encoding that needs the MonthPicker/WeekPicker conversion. Added a comment to make this clearer.

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 3]
```diff
@@ -32,12 +38,15 @@ enum DefaultToNow {
         YES,
         NO
     }
-    private static Date parseDate(final SimpleDateFormat formatter, final String value, final DefaultToNow defaultToNow) {
+
+    private static Date parseDate(final SimpleDateFormat formatter, final String value,
+                                  final DefaultToNow defaultToNow) {
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: fixed thanks

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 4]
```diff
@@ -32,12 +38,15 @@ enum DefaultToNow {
...
         try {
             if (value != null && !value.isEmpty()) {
                 return formatter.parse(value);
             }
         } catch (final ParseException e) {
+            // fall through
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: removed.

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 5]
```diff
@@ -82,12 +91,19 @@ protected void initialize(Context aContext) {
         mOkButton = findViewById(R.id.positiveButton);
 
         final SimpleDateFormat formatter = new SimpleDateFormat(format, Locale.ROOT);
+        formatter.setTimeZone(TimeZone.getTimeZone("UTC"));
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: Good call, removed the forced UTC. It'll use the device timezone now.

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 6]
```diff
@@ -152,7 +216,16 @@ protected void initialize(Context aContext) {
                     cal.set(Calendar.HOUR_OF_DAY, timePicker.getHour());
                     cal.set(Calendar.MINUTE, timePicker.getMinute());
                 }
-                ((DateTimePromptDelegate) mPromptDelegate).confirm(formatter.format(cal.getTime()));
+
+                // Round to nearest step boundary if step is set
+                final String confirmed;
+                if (!Double.isNaN(stepValue) && stepValue > 0) {
+                    confirmed = formatter.format(snapToStep(cal.getTime(), stepValue,
+                            minDate, mPrompt.type()));
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: done

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 7]
```diff
@@ -163,6 +236,56 @@ protected void initialize(Context aContext) {
...
+    private static double parseStep(@Nullable String stepStr) {
+        if (stepStr == null || stepStr.isEmpty() || "any".equalsIgnoreCase(stepStr)) {
+            return Double.NaN;
+        }
+        try { return Double.parseDouble(stepStr); } catch (NumberFormatException e) { return Double.NaN; }
+    }
+
+    // Parse suggestion strings to Dates, skipping entries that fail.
+    private static List<Date> buildSuggestions(SimpleDateFormat formatter,
+                                               @Nullable String[] listValues) {
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: fixed thanks

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 8]
```diff
@@ -163,6 +236,56 @@ protected void initialize(Context aContext) {
...
+    // HTML attribute unit: days for date, weeks for week, seconds for time.
+    // Month is skipped - non-uniform month lengths make ms arithmetic unreliable.
+    private static Date snapToStep(Date selected, double stepValue,
+                                   @Nullable Date minDate, int type) {
+        final long stepMs;
+        switch (type) {
+            case WSession.PromptDelegate.DateTimePrompt.Type.MONTH:
+                return selected;
+            case WSession.PromptDelegate.DateTimePrompt.Type.WEEK:
+                stepMs = Math.round(stepValue * 7L * 24L * 60L * 60L * 1000L);
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: Yep, same constants as above - MS_PER_WEEK, MS_PER_DAY, MS_PER_SECOND.

File: app/src/main/res/layout/prompt_datetime.xml [hunk 9]
```diff
@@ -45,19 +45,32 @@
                         android:layout_height="wrap_content"
                         android:background="@drawable/prompt_background"
                         android:layout_margin="20dp"
-                        android:orientation="horizontal">
+                        android:orientation="vertical">
```
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/main/res/layout/prompt_datetime.xml:48]: Needed to stack the suggestions spinner above the date/time pickers. Changed layout_picker to vertical and wrapped the actual pickers in a nested horizontal LinearLayout so they still sit side by side.

**Comment by svillar** [comment]
@ojuschugh1 could you fix the conflicts?

Also do you have any demo we could use to test it?

**Comment by ojuschugh1** [comment]
> @ojuschugh1 could you fix the conflicts?
> 
> Also do you have any demo we could use to test it?




Rebased on latest main, conflicts are fixed.

For a demo, here's a small HTML page you can drop into a gist or any static host: it covers `step` and `list` across date, time, datetime-local, week, and month inputs.

```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>DateTime step &amp; list demo</title>
    <style>
        body { font-family: sans-serif; max-width: 640px; margin: 2em auto; }
        label { display: block; margin-top: 1.2em; font-weight: bold; }
        input { font-size: 1rem; padding: 4px; }
        small { color: #666; display: block; margin-top: 2px; }
        h2 { margin-top: 2em; border-bottom: 1px solid #ccc; padding-bottom: 4px; }
        #output { margin-top: 2em; padding: 1em; background: #f4f4f4; white-space: pre-wrap; font-family: monospace; }
    </style>
</head>
<body>
    <h1>DateTime step &amp; datalist test</h1>
    <p>Each input below exercises a different combination of <code>step</code>
       and <code>list</code> attributes. Pick a value and check the result at
       the bottom.</p>

    <h2>date</h2>

    <label>step=7 (weekly increments)
        <input type="date" id="date-step" step="7" min="2026-04-20">
    </label>
    <small>min=2026-04-20 (a Monday), step=7 &rarr; should snap to Mondays</small>

    <label>datalist suggestions
        <input type="date" id="date-list" list="date-options">
    </label>
    <datalist id="date-options">
        <option value="2026-06-15">
        <option value="2026-09-01">
        <option value="2026-12-25">
    </datalist>
    <small>Suggestions: Jun 15, Sep 1, Dec 25</small>

    <h2>time</h2>

    <label>step=900 (15-min intervals)
        <input type="time" id="time-step" step="900" min="09:00" max="17:00">
    </label>
    <small>min=09:00, max=17:00, step=900</small>

    <label>datalist suggestions
        <input type="time" id="time-list" list="time-options">
    </label>
    <datalist id="time-options">
        <option value="08:30">
        <option value="12:00">
        <option value="18:45">
    </datalist>
    <small>Suggestions: 08:30, 12:00, 18:45</small>

    <h2>datetime-local</h2>

    <label>step=3600 (hourly)
        <input type="datetime-local" id="dtl-step" step="3600" min="2026-05-01T08:00">
    </label>
    <small>min=2026-05-01T08:00, step=3600</small>

    <h2>week</h2>

    <label>step=2 (every other week)
        <input type="week" id="week-step" step="2" min="2026-W01">
    </label>
    <small>min=2026-W01, step=2</small>

    <h2>month</h2>

    <label>step=3 (quarterly)
        <input type="month" id="month-step" step="3" min="2026-01">
    </label>
    <small>min=2026-01, step=3 (month step is accepted but not snapped &mdash; non-uniform month lengths)</small>

    <h2>Result</h2>
    <div id="output">Pick a value above...</div>

    <script>
        const out = document.getElementById('output');
        document.querySelectorAll('input').forEach(input => {
            input.addEventListener('change', () => {
                out.textContent = input.id + ' = ' + input.value;
            });
        });
    </script>
</body>
</html>
```

Things to check:

- **date/step=7**: confirming any date should snap to a Monday (base is 2026-04-20).
- **time/step=900** with min/max: picks should snap to :00/:15/:30/:45 and clamp inside 09:00–17:00.
- **datalist on Chromium**: a spinner appears above the pickers with the options; selecting one updates the pickers.
- **datalist on Gecko**: no spinner (GeckoView doesn't expose datalist), the picker still works normally. This is the graceful fallback mentioned in the PR description.
- **month/step**: accepted without error but not snapped (month lengths aren't uniform).

**Comment by svillar** [comment]
@ojuschugh1 please take a look at the issues reported by copilot.

**Comment by ojuschugh1** [comment]
> @ojuschugh1 please take a look at the issues reported by copilot.

Sure, i waiting your comment to check explicitly which issues are geniune that needed to be resolved , since you know llms keep on commenting out even on small grammar mistakes as well lol.

No issue will fix these issues and comment on them  as well. Thanks

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 10]
```diff
@@ -88,6 +94,12 @@ protected void initialize(Context aContext) {
         final Calendar cal = formatter.getCalendar();
         cal.setTime(date);
```
[the remark this answers is not in the document]
Comment (svillar) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java:95]: not sure this is really an issue

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 11]
```diff
@@ -179,7 +239,15 @@ protected void initialize(Context aContext) {
                     cal.set(Calendar.HOUR_OF_DAY, timePicker.getHour());
                     cal.set(Calendar.MINUTE, timePicker.getMinute());
                 }
-                ((DateTimePromptDelegate) mPromptDelegate).confirm(formatter.format(cal.getTime()));
+
+                // Round to nearest step boundary if step is set
+                final String confirmed;
+                if (!Double.isNaN(stepValue) && stepValue > 0) {
+                    confirmed = formatter.format(snapToStep(cal.getTime(), stepValue, minDate, mPrompt.type()));
```
[the remark this answers is not in the document]
Comment (svillar) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: seems legit

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java [hunk 12]
```diff
@@ -242,6 +310,58 @@ private static boolean isOnBoundaryDay(Calendar boundaryCal, DatePicker datePick
...
+    private static final long MS_PER_SECOND = 1000L;
+    private static final long MS_PER_DAY = 24L * 60L * 60L * MS_PER_SECOND;
+    private static final long MS_PER_WEEK = 7L * MS_PER_DAY;
+
+    private static Date snapToStep(Date selected, double stepValue, @Nullable Date minDate, int type) {
+        final long stepMs;
+        switch (type) {
+            case WSession.PromptDelegate.DateTimePrompt.Type.MONTH:
+                return selected;
+            case WSession.PromptDelegate.DateTimePrompt.Type.WEEK:
```
[the remark this answers is not in the document]
Comment (svillar) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java:348]: there is no type=month AFAIK

**Review by svillar — CHANGES_REQUESTED** [review]
Yeah you're right, I did the review but forgot to publish it 🤦

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java]: Fixed - snapToStep now takes maxDate as well and clamps the result to [min, max] before returning. Added a couple of tests for the boundary cases.

File: app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (ojuschugh1) [reply app/src/common/shared/com/igalia/wolvic/ui/widgets/prompts/DateTimePromptWidget.java:348]: Yeah fair, left it as a no-op. If it ever gets hit it just returns the selected value unchanged.

**Review by svillar — APPROVED** [review]
Seems ready to go

Thanks for the effort!
